File: estructurales/adapter/AdapterExample.java
interface Printer {
    void print(String message);
}

// Clase existente con una interfaz distinta a la esperada
class LegacyPrinter {
    public void printText(String text) {
        System.out.println("LegacyPrinter: " + text);
    }
}

// Adapter: implementa la interfaz esperada y delega al LegacyPrinter
class PrinterAdapter implements Printer {
    private final LegacyPrinter legacy;

    public PrinterAdapter(LegacyPrinter legacy) {
        this.legacy = legacy;
    }

    @Override
    public void print(String message) {
        legacy.printText(message);
    }
}

public class AdapterExample {
    public static void main(String[] args) {
        LegacyPrinter legacy = new LegacyPrinter();
        Printer printer = new PrinterAdapter(legacy);

        printer.print("Hola desde Adapter (Java)");
    }
}

File: estructurales/adapter/AdapterExample.py
class LegacyPrinter:
    # Clase existente con una interfaz "vieja"
    def print_text(self, text: str) -> None:
        print(f"LegacyPrinter: {text}")


class Printer:
    # Interfaz esperada por el cliente
    def print(self, message: str) -> None:
        raise NotImplementedError


class PrinterAdapter(Printer):
    # Adapter: hace que LegacyPrinter funcione como Printer
    def __init__(self, legacy_printer: LegacyPrinter) -> None:
        self.legacy = legacy_printer

    def print(self, message: str) -> None:
        self.legacy.print_text(message)


if __name__ == "__main__":
    legacy = LegacyPrinter()
    printer: Printer = PrinterAdapter(legacy)

    printer.print("Hola desde Adapter (Python)")